platforms/cuda/src/CudaTorchKernels.cpp

```diff
@@ -92,96 +92,157 @@ void CudaCalcTorchForceKernel::initialize(const System& system, const TorchForce
-        if (!outputsForces) {
+        // Compute force by backpropagating the PyTorch model
+        if (includeForces) {
             energyTensor.backward();
-            forceTensor = posTensor.grad();
-        }
-
-        // Get a pointer to the computed forces
-        void* forceData;
-        if (cu.getUseDoublePrecision()) {
-            if (!(forceTensor.dtype() == torch::kFloat64)) // TODO: simplify the logic when support for PyTorch 1.7 is dropped
-                forceTensor = forceTensor.to(torch::kFloat64);
-            forceData = forceTensor.data_ptr<double>();
-        }
-        else {
-            if (!(forceTensor.dtype() == torch::kFloat32)) // TODO: simplify the logic when support for PyTorch 1.7 is dropped
-                forceTensor = forceTensor.to(torch::kFloat32);
-            forceData = forceTensor.data_ptr<float>();
+            forceTensor = posTensor.grad().clone();
+	    // Zero the gradient to avoid accumulating it
+            posTensor.grad().zero_();
         }
-        CHECK_RESULT(cuCtxSynchronize(), "Failed to synchronize the CUDA context"); // Synchronize before switching to the OpenMM context
+    }
+}
 
-        // Add the computed forces to the total atomic forces
-        {
-            ContextSelector selector(cu); // Switch to the OpenMM context
-            int paddedNumAtoms = cu.getPaddedNumAtoms();
-            int forceSign = (outputsForces ? 1 : -1);
-            void* forceArgs[] = {&forceData, &cu.getForce().getDevicePointer(), &cu.getAtomIndexArray().getDevicePointer(), &numParticles, &paddedNumAtoms, &forceSign};
-            cu.executeKernel(addForcesKernel, forceArgs, numParticles);
-            CHECK_RESULT(cuCtxSynchronize(), "Failed to synchronize the CUDA context"); // Synchronize before switching to the PyTorch context
+double CudaCalcTorchForceKernel::execute(ContextImpl& context, bool includeForces, bool includeEnergy) {
+    // Push to the PyTorch context
+    CHECK_RESULT(cuCtxPushCurrent(primaryContext), "Failed to push the CUDA context");
+    auto inputs = prepareTorchInputs(context);
+    if (!useGraphs) {
+        execute_graph(outputsForces, includeForces, module, inputs, posTensor, energyTensor, forceTensor);
+    } else {
+#if CUDA_GRAPHS_SUPPORTED
+        const auto stream = c10::cuda::getStreamFromPool(false, posTensor.get_device());
+        const c10::cuda::CUDAStreamGuard guard(stream);
+        // Record graph if not already done
+        bool is_graph_captured = false;
+        if (graphs.find(includeForces) == graphs.end()) {
+            // Warmup the graph workload before capturing.  This first
+            // run  before  capture sets  up  allocations  so that  no
+            // allocations are  needed after.  Pytorch's  allocator is
+            // stream  capture-aware and,  after warmup,  will provide
+            // record static pointers and shapes during capture.
+            for (int i = 0; i < 10; i++)
```

Comment:
There is no reason at all. The pytorch docs suggest running the model "a few times" before graphing. To me only one, maybe two should suffice. But just in case...
The tests pass with just one warmup execution. Would you use just one?

Reply:
Maybe we should expose it as an option with a reasonable default. Some models might do auto-tuning and require a specific number of executions.

Reply:
Defaulted warmup to 1 step. Added a property CUDAGraphWarmupSteps to control this.
AFAIK the model has to be run at least once before capture, though. Right now, the call to execute that captures the graph will effectively run the model at least twice.

Reply:
Great!